### ISSUE TITLE
clangd unable to parse some header files due to exponential memory usage causing OOM

### ISSUE THREAD

**Issue by drvink** [clangd/clangd#1384]
clangd consumes all memory on my system (~24GiB) before being OOM killed while attempting to parse [this attached header file][hdr] (rename it to `mathlink.h` because GitHub is stupid and won't let me upload a file with a `.h` extension on a web site about programming).

[hdr]: <https://github.com/clangd/clangd/files/10046587/mathlink.txt>

Presumably it's stuck in a loop that would never terminate; or it might if given more memory, but would require an astronomical amount to succeed--notice, however, that the header contents are hardly exotic: just preprocessor junk and C declarations.

Repro with `systemd-run --user -p MemoryMax=8G -S`, then in the shell, run `clangd --log=verbose --check=mathlink.h` (you don't **have** to use `systemd-run`, but beware of systemd/systemd#25376 if you don't!)

**Logs**

[`clangd-stderr.txt`](https://github.com/clangd/clangd/files/10046563/clangd-stderr.txt)

**System information**

Output of `clangd --version`:

```
clangd version 15.0.3 (https://github.com/llvm/llvm-project 4a2c05b05ed07f1f620e94f6524a8b4b2760a0b1)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu
```

Editor/LSP plugin:

Emacs 28.1 + lsp-mode

Operating system:

```
Linux arch-vm 6.0.7-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 03 Nov 2022 18:01:58 +0000 x86_64 GNU/Linux
```

**Comment by drvink** [comment]
This is still broken on clangd 16.0.2.

```
clangd version 16.0.2 (https://github.com/llvm/llvm-project 18ddebe1a1a9bde349441631365f0472e9693520)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu
```

There's a newer version of the header file I attached in the initial post that parses successfully, but the diff is a bit large to be able to wager a guess at exactly what causes the problem in the original file. I've attached it anyway in case it's useful: [mathlink2.txt](https://github.com/clangd/clangd/files/11648135/mathlink2.txt)

**Comment by HighCommander4** [comment]
I tried this locally, and attached a debugger during the hang to get a stack trace.

The stack trace shows us being in `clang::format::guessLanguage()`, which makes this another case of https://github.com/llvm/llvm-project/issues/60151 (which has been reported for clangd with other libraries such as miniaudio.h, e.g. in https://github.com/clangd/clangd/issues/719).

**Comment by drvink** [comment]
Thanks for looking into this. It's almost definitely the same root cause as the bug you linked, since renaming the file makes the issue go away here too.

I started following the stack trace down from [guessLanguage()](https://github.com/llvm/llvm-project/blob/7cbf1a2591520c2491aa35339f227775f4d3adf6/clang/lib/Format/Format.cpp#L3704) and arrived at the extremely dodgy-looking [UnwrappedLineParser::parseLevel()](https://github.com/llvm/llvm-project/blob/7cbf1a2591520c2491aa35339f227775f4d3adf6/clang/lib/Format/UnwrappedLineParser.cpp#L494)--will you be surprised if I tell you it accumulates tokens in `do ... while (!feof())` loop? Given llvm/llvm-project#78263 has a minimized repro and my original file exercises the exact same combination of tokens/state, this is looking promising. See also llvm/llvm-project#64328.

**Comment by HighCommander4** [comment]
> That said, after some experimentation with the file I provided, it seems less likely that it's an issue of outright non-termination, but rather that the algorithm is susceptible to exponential blowup (see [llvm/llvm-project#64328](https://github.com/llvm/llvm-project/issues/64328)).

Yeah, Sam came to the same conclusion in https://github.com/llvm/llvm-project/issues/48863#issuecomment-981039782.

I tried to come up with a way to work around this in clangd (see https://github.com/clangd/clangd/issues/719#issuecomment-1789776339 and subsequent comments) but we don't seem to have great options.

I'm now thinking a better avenue would be to add some sort of guard against algorithmic blowup into the relevant clang-format code (whether a depth limit, or a maximum iteration count, or somethig like that), and bail if we hit that limit, possibly having `guessLanguage()` produce a suboptimal result which is still much better than OOM-ing.

**Comment by drvink** [comment]
I agree with your comments.

Frankly, I'd remove the silly heuristic entirely--it's slow, failure-prone, and does a disservice to the majority of users in favor of offering dubious benefit to a much smaller minority, who would only have to add overrides in order to view `.h` objc headers that hadn't already been inferred by clangd as objc (easy to track, since objc uses `#import`). Just add a handful of default paths that are typically known to contain objc headers or whatever--`/usr/include/objc`, wherever Apple's framework paths look like these days, etc.

**Comment by HighCommander4** [comment]
> I'm now thinking a better avenue would be to add some sort of guard against algorithmic blowup into the relevant clang-format code (whether a depth limit, or a maximum iteration count, or somethig like that), and bail if we hit that limit, possibly having `guessLanguage()` produce a suboptimal/incorrect result which is still much better than OOM-ing.

I implemented this in https://github.com/llvm/llvm-project/pull/78925, and confirmed that it addresses the OOM in `guessLanguage()` in the various libraries for which this bug was reported (mathlink.h, miniaudio.h, a couple of others).

> Frankly, I'd remove the silly heuristic entirely--it's slow, failure-prone, and does a disservice to the majority of users in favor of offering dubious benefit to a much smaller minority, who would only have to add overrides in order to view `.h` objc headers that hadn't already been inferred by clangd as objc (easy to track, since objc uses `#import`). Just add a handful of default paths that are typically known to contain objc headers or whatever--`/usr/include/objc`, wherever Apple's framework paths look like these days, etc.

I didn't pursue this because currently there isn't a mechanism for clangd to tell clangd-format the file type, and in any case as discussed [here](https://github.com/clangd/clangd/issues/719#issuecomment-1790284919), a discrepancy between "filetype detected when invoked via clangd" and "filetype detected when invoked on the command line" would be problematic. One could conceivably add a way to configure file type overrides at the clang-format level (e.g. in a `.clang-format` file); if you or someone would like to pursue that and remove `guessLanguage()` altogether, please feel free.

**Comment by drvink** [comment]
> I implemented this in [llvm/llvm-project#78925](https://github.com/llvm/llvm-project/pull/78925), and confirmed that it addresses the OOM in `guessLanguage()` in the various libraries for which this bug was reported (mathlink.h, miniaudio.h, a couple of others).

Thank you so much!!

> I didn't pursue this because currently there isn't a mechanism for clangd to tell clang-format the file type, and in any case as discussed [here](https://github.com/clangd/clangd/issues/719#issuecomment-1790284919), a discrepancy between "filetype detected when invoked via clangd" and "filetype detected when invoked on the command line" would be problematic. One could conceivably add a way to configure file type overrides at the clang-format level (e.g. in a `.clang-format` file); if you or someone would like to pursue that and remove `guessLanguage()` altogether, please feel free.

I see; makes sense.
